src/Features/Core/Portable/ExtractClass/ExtractClassWithDialogCodeAction.cs

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+
+        private async Task<Solution> PullMembersUpAsync(
+            Solution solution,
+            INamedTypeSymbol newType,
+            AnnotatedSymbolMapping symbolMapping,
+            ImmutableArray<ExtractClassMemberAnalysisResult> memberAnalysisResults,
+            CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<(ISymbol member, bool makeAbstract)>.GetInstance(out var pullMembersBuilder);
+
+            var remainingResults = new List<ExtractClassMemberAnalysisResult>(memberAnalysisResults);
+
+            foreach (var documentId in symbolMapping.DocumentIds)
+            {
+                if (remainingResults.Count == 0)
+                {
+                    // All symbols have been taken care of
+                    break;
+                }
+
+                var document = solution.GetRequiredDocument(documentId);
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var typeDeclaration = root.GetAnnotatedNodes(symbolMapping.TypeNodeAnnotation).SingleOrDefault();
+                if (typeDeclaration == null)
+                {
+                    continue;
+                }
+
+                foreach (var memberAnalysis in remainingResults.ToArray())
+                {
+                    var annotation = symbolMapping.SymbolToDeclarationAnnotationMap[memberAnalysis.Member];
+
+                    var nodeOrToken = typeDeclaration.GetAnnotatedNodesAndTokens(annotation).SingleOrDefault();
+                    var node = nodeOrToken.IsNode
+                        ? nodeOrToken.AsNode()
+                        : nodeOrToken.AsToken().Parent;
+
+                    if (node is null)
+                    {
+                        continue;
+                    }
+
+                    var currentSymbol = semanticModel.GetDeclaredSymbol(node, cancellationToken);
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    pullMembersBuilder.Add((currentSymbol, memberAnalysis.MakeAbstract));
+                    remainingResults.Remove(memberAnalysis);
+                }
+            }
+
+            Contract.ThrowIfFalse(remainingResults.Count == 0);
+
+            var pullMemberUpOptions = PullMembersUpOptionsBuilder.BuildPullMembersUpOptions(newType, pullMembersBuilder.ToImmutable());
+            var updatedOriginalDocument = solution.GetRequiredDocument(_document.Id);
+
+            return await MembersPuller.PullMembersUpAsync(updatedOriginalDocument, pullMemberUpOptions, cancellationToken).ConfigureAwait(false);
```

Comment:
baed on how much sharing is goign on, i'm strongly leaning to this going in the PullMembers namespace and just being a different code action that is offered depending on if the tyep has an existing base type or not.  it really feels like less of a "separate feature" and more of a "how do we handle this particular case of this feature".

Reply:
https://github.com/dotnet/roslyn/pull/46105#issuecomment-663774118